metric_learn/oasis.py

```diff
@@ -0,0 +1,375 @@
+                                       random=self.random_sampling,
+                                       random_state=self.random_state,
+                                       custom=self.custom_order)
+    i = 0
+    while i < self.n_iter:
+        current_triplet = X[triplets[self.indices[i]]]
+        loss = self._loss(current_triplet)
+        vi = self._vi_matrix(current_triplet)
+        fs = np.linalg.norm(vi, ord='fro') ** 2
+        # Global GD or Adjust to tuple
+        tau_i = np.minimum(self.c, loss / fs)
+
+        # Update components
+        self.components_ = np.add(self.components_, tau_i * vi)
+        i = i + 1
+
+    return self
+
+  def partial_fit(self, new_triplets, n_iter, shuffle=True,
+                  random_sampling=False, custom_order=None):
+    """
+    Reuse previous fit, and feed the algorithm with new triplets.
+    A new n_iter can be set for these new triplets.
+
+    Parameters
+    ----------
+    new_ triplets : (n x 3 x d) array of samples
+
+    n_iter: int (default = n_triplets)
+      Number of iterations. When n_iter < n_triplets, a random sampling
+      takes place without repetition, but preserving the original order.
+      When n_iter = n_triplets, all triplets are included with the
+      original order. When n_iter > n_triplets, each triplet is included
+      at least floor(n_iter/n_triplets) times, while some may have one
+      more apparition at most. The order is preserved as well.
+
+    shuffle: bool (default = True)
+      Whether the triplets should be shuffled after the sampling process.
+      If n_iter > n_triplets, then the suffle happends during the sampling
+      and at the end.
+
+    random_sampling: bool (default = False)
+      If enabled, the algorithm will sample n_iter triplets from
+      the input. This sample can contain duplicates. It does not
+      matter if n_iter is lower, equal or greater than the number
+      of triplets. The sampling uses uniform distribution.
+
+    custom_order : array-like, optinal (default = None)
+      User's custom order of triplets to feed oasis. Might be useful when
+      trying to put a bias in the resulting similarity matrix.
+    """
+    self.n_iter = n_iter
+    self.fit(new_triplets, shuffle=shuffle, random_sampling=random_sampling,
+             custom_order=custom_order)
+
+  def _loss(self, triplet):
+    """
+    Loss function in a triplet
+    """
+    S = self.pair_similarity([[triplet[0], triplet[1]],
+                              [triplet[0], triplet[2]]])
+    return np.maximum(0, 1 - S[0] + S[1])
+
+  def _vi_matrix(self, triplet):
+    """
+    Computes V_i, the gradient matrix in a triplet
+    """
+    # (pi+ - pi-)
+    diff = np.subtract(triplet[1], triplet[2])  # Shape (, d)
+    result = []
+
+    # For each scalar in first triplet, multiply by the diff of pi+ and pi-
+    for v in triplet[0]:
+        result.append(v * diff)
+
+    return np.array(result)  # Shape (d, d)
```

Comment:
This part is correct, just checked manually, but I've found a better way to compute it. This is equivalent to:
```python
return np.outer(triplet[0], np.subtract(triplet[1], triplet[2]))
```
For d=100, and executing this function n=1.000.000 times, I get the following timings:
```
Current solution: 139.7970213389999 [s]
Using numpy outer: 13.34107805200074 [s]
```
I've changed it for this optimized approach